DPGAnalysis/Skims/src/IsoPhotonEBSelector.cc

```diff
@@ -0,0 +1,141 @@
+
+#include "DataFormats/EgammaCandidates/interface/Photon.h"
+#include "DataFormats/EgammaCandidates/interface/PhotonFwd.h"
+#include "DataFormats/GsfTrackReco/interface/GsfTrack.h"
+#include "DataFormats/JetReco/interface/PFJetCollection.h"
+#include "CommonTools/UtilAlgos/interface/SingleObjectSelector.h"
+//
+#include "FWCore/Utilities/interface/InputTag.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "CommonTools/UtilAlgos/interface/SingleObjectSelector.h"
+#include "DataFormats/Common/interface/View.h"
+
+using namespace std;
+using namespace reco;
+namespace edm {
+  class EventSetup;
+}
+
+class IsoPhotonEBSelector {
+public:
+  IsoPhotonEBSelector(const edm::ParameterSet&, edm::ConsumesCollector& iC);
+  bool operator()(const reco::Photon&) const;
+  void newEvent(const edm::Event&, const edm::EventSetup&);
+  const float getEffectiveArea(float eta) const;
+  void printEffectiveAreas() const;
+
+  edm::EDGetTokenT<double> theRhoToken;
+  edm::EDGetTokenT<reco::PhotonCollection> thePhotonToken;
+  edm::Handle<double> _rhoHandle;
+
+  std::vector<double> absEtaMin_;            // low limit of the eta range
+  std::vector<double> absEtaMax_;            // upper limit of the eta range
+  std::vector<double> effectiveAreaValues_;  // effective area for this eta range
+
+  edm::ParameterSet phIDWP;
+
+  vector<double> sigmaIEtaIEtaCut;
+  vector<double> hOverECut;
+  vector<double> relCombIso;
+};
+
+void IsoPhotonEBSelector::printEffectiveAreas() const {
+  printf("  eta_min   eta_max    effective area\n");
+  uint nEtaBins = absEtaMin_.size();
+  for (uint iEta = 0; iEta < nEtaBins; iEta++) {
+    printf("  %8.4f    %8.4f   %8.5f\n", absEtaMin_[iEta], absEtaMax_[iEta], effectiveAreaValues_[iEta]);
+  }
+}
+const float IsoPhotonEBSelector::getEffectiveArea(float eta) const {
+  float effArea = 0;
+  uint nEtaBins = absEtaMin_.size();
+  for (uint iEta = 0; iEta < nEtaBins; iEta++) {
+    if (std::abs(eta) >= absEtaMin_[iEta] && std::abs(eta) < absEtaMax_[iEta]) {
+      effArea = effectiveAreaValues_[iEta];
+      break;
+    }
+  }
+
+  return effArea;
+}
+
+IsoPhotonEBSelector::IsoPhotonEBSelector(const edm::ParameterSet& cfg, edm::ConsumesCollector& iC)
+    : theRhoToken(iC.consumes<double>(cfg.getParameter<edm::InputTag>("rho"))) {
+  absEtaMin_ = cfg.getParameter<std::vector<double> >("absEtaMin");
+  absEtaMax_ = cfg.getParameter<std::vector<double> >("absEtaMax");
+  effectiveAreaValues_ = cfg.getParameter<std::vector<double> >("effectiveAreaValues");
+  //printEffectiveAreas();
+  phIDWP = cfg.getParameter<edm::ParameterSet>("phID");
+
+  sigmaIEtaIEtaCut = phIDWP.getParameter<std::vector<double> >("full5x5_sigmaIEtaIEtaCut");
+  hOverECut = phIDWP.getParameter<std::vector<double> >("hOverECut");
+  relCombIso = phIDWP.getParameter<std::vector<double> >("relCombIsolationWithEACut");
+}
+
+void IsoPhotonEBSelector::newEvent(const edm::Event& ev, const edm::EventSetup&) {
```

Comment:
I miss to see where this `newEvent` method is ever used.
Did you check whether the `_rhoHandle` was ever valid?
By the way, there is no need to make `_rhoHandle` a class member: it can be defined and accessed before [L127](https://github.com/cms-sw/cmssw/pull/42150/files#diff-84f9171bab0d7f088d5b47b20c6cd871b5e3a529b177103ff68329b416edae97R127) from the token

Reply:
Yes I explicitly checked that `_rhoHandle` is valid and that the product makes sense. My understanding is the `newEvent` is used when calling 
https://github.com/cms-sw/cmssw/pull/42150/files#diff-84f9171bab0d7f088d5b47b20c6cd871b5e3a529b177103ff68329b416edae97R130
more precisely here: 
https://github.com/cms-sw/cmssw/blob/master/CommonTools/UtilAlgos/interface/EventSetupInitTrait.h#L32-L34